src/codeu/chat/server/Server.java

```diff
@@ -276,11 +305,80 @@ public void run() {
         final ConversationHeader conversation = view.findConversation(conversationId);
         final Message message = view.findMessage(messageId);
         relay.write(id,
-                    secret,
-                    relay.pack(user.id, user.name, user.creation),
-                    relay.pack(conversation.id, conversation.title, conversation.creation),
-                    relay.pack(message.id, message.content, message.creation));
+                secret,
+                relay.pack(user.id, user.name, user.creation),
+                relay.pack(conversation.id, conversation.title, conversation.creation),
+                relay.pack(message.id, message.content, message.creation));
       }
     };
   }
+
+  private void process(String line) throws IOException {
+    //turns log string into an array of words
+    String[] splitLog = line.split(" ");
+
+    //uuid and time occupy the same spot in each format so they have been extracted ahead of time
+    Uuid uuid = Uuid.parse(splitLog[1]);
+    long time = Long.parseLong(splitLog[3]);
+
+    switch (splitLog[0]) {
+      case "ADD-USER":
+        String name = splitLog[2];
+        controller.newUser(uuid, name, Time.fromMs(time));
+        break;
+      case "ADD-CONVERSATION":
+        String title = splitLog[2];
+        Uuid owner = Uuid.parse(splitLog[4]);
+        controller.newConversation(uuid, title, owner, Time.fromMs(time));
+        break;
+      case "ADD-MESSAGE":
+        String message = splitLog[2];
+        Uuid conversationUuid = Uuid.parse(splitLog[4]);
+        Uuid senderUuid = Uuid.parse(splitLog[5]);
+        controller.newMessage(uuid, senderUuid, conversationUuid, message, Time.fromMs(time));
+        break;
+    }
+  }
+
+  public void restore(File persistentPath) {
```

Comment:
again, just pass the file path to the controller and let the controller handle the logic. see comment below for more context.